### ISSUE TITLE
Strange behavior with Watch statement

### ISSUE THREAD

**Issue by mcscope** [jamesls/fakeredis#72]
I think I've discovered a discrepancy between redis's watch statement implementation and fakeredis's mocking of that implementation. 
Code that runs fine on actual redis throws a watch error every time when using fakeredis. 

I reduced the code down into the minimal example, below.

```
key = 'testkey'
fake_redis = fakeredis.FakeStrictRedis()
redis = RedisProxy() # however you connect to your redis
with fake_redis.pipeline(transaction=True) as pl:
    pl.watch(key)
    pl.hset(key, 'blah1', "blah2")

    pl.execute()
```

if you use redis.pipeline, this code works fine, if you use fake_redis.pipeline, you get the following

```
In [8]:     pl.execute()
---------------------------------------------------------------------------
WatchError                                Traceback (most recent call last)
<ipython-input-8-622a4d39c733> in <module>()
----> 1 pl.execute()

/usr/local/Cellar/python/2.7.8_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/fakeredis-0.5.1-py2.7.egg/fakeredis.pyc in execute(self)
   1247                     'Watched key%s %s changed' % (
   1248                         '' if len(mismatches) == 1 else
-> 1249                         's', ', '.join(k for (k, _, _) in mismatches)))
   1250         ret = [getattr(self.owner, name)(*args, **kwargs)
   1251                for name, args, kwargs in self.commands]

WatchError: Watched key testkey changed

```

A watch shouldn't throw an error if it's modified by the same connection that set the watch - it should only error if it's modified by another connection, to prevent race conditions.

**Comment by ketralnis** [comment]
In particular, fakeredis' `watch` implementation fails redis' own canonical use of watch (http://redis.io/topics/transactions#optimistic-locking-using-check-and-set) which is to watch a key and increment it

**Comment by mcscope** [comment]
If we write a patch for this, might you accept it?  It's causing us trouble

**Comment by jamesls** [comment]
@mcscope definitely.  Do keep in mind the [contributing guidelines](https://github.com/jamesls/fakeredis/blob/master/CONTRIBUTING.rst), namely that there has to be tests for the code change.

**Comment by ketralnis** [comment]
Here's the bug as I understand it so far:

When you `watch` a key, the `FakePipeline` makes a copy of it and keeps it around. Before the pipeline is executed, those keys are checked in the parent store. If they aren't equal, the pipeline fails.

But when you're watching a key, the pipeline goes into `is_immediate` mode. In this mode, the pipeline is bypassed entirely and changes are committed directly to the store:

        def meth(*args, **kwargs):
            if self.is_immediate:
                # Special mode during watch_multi sequence.
                return getattr(self.owner, name)(*args, **kwargs)

            self.commands.append((name, args, kwargs))
            return self

Then when the pipeline is later executed, the `watching` keys are different from the ones in the store precisely because we changed them right in the store instead of buffering them up.

This means that the `watching` dictionary can't have ever worked, because the only method that sets it also sets `is_immediate` and these are incompatible

**Comment by bmerry** [comment]
> Is there any progress on this?

I'm not aware of anyone working on it. fakeredis is on very limited maintenance: the fastest way to get something fixed is to submit a pull request.

**Comment by bmerry** [comment]
This is fixed in 0.14.0, at least for the sample code posted. However, note that the sample is **not** transactional, because it does not use `pipeline.multi`.
